apm-agent-plugins/apm-process-plugin/src/test/java/co/elastic/apm/agent/process/ProcessHelperTest.java

```diff
@@ -0,0 +1,161 @@
+
+import javax.annotation.Nullable;
+
+import java.nio.file.Paths;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class ProcessHelperTest extends AbstractInstrumentationTest {
+
+    // implementation note:
+    //
+    // Testing instrumentation of classes loaded by bootstrap classloader can't be done with regular unit tests
+    // as the agent is loaded in the application/system classloader when they are run.
+    //
+    // Hence, in order to maximize test coverage we thoroughly test helper implementation where most of the actual code
+    // of this instrumentation is. Also, integration test cover this feature for the general case with a packaged
+    // agent and thus they don't have such limitation
+
+    @Nullable
+    private Transaction transaction = null;
+
+    private WeakConcurrentMap<Process, Span> storageMap;
+    private ProcessHelper helper;
+
+    @BeforeEach
+    void before() {
+        transaction = new Transaction(tracer);
+        TransactionUtils.fillTransaction(transaction);
+
+        storageMap = DataStructures.createWeakConcurrentMapWithCleanerThread();
+        helper = new ProcessHelper(storageMap);
+    }
+
+    @Test
+    void checkSpanNaming() {
+        Process process = mock(Process.class);
+
+        String binaryName = "hello";
+        String programName = Paths.get("bin", binaryName).toAbsolutePath().toString();
+
+        helper.doStartProcess(transaction, process, programName);
+
+        helper.doWaitForEnd(process);
+
+        assertThat(reporter.getSpans()).hasSize(1);
+        Span span = reporter.getSpans().get(0);
+
+        assertThat(span.getNameAsString()).isEqualTo(binaryName);
+        assertThat(span.getType()).isEqualTo("process");
+        assertThat(span.getSubtype()).isEqualTo(binaryName);
+        assertThat(span.getAction()).isEqualTo("execute");
+    }
+
+    @Test
+    void startTwiceShouldIgnore() {
+        Process process = mock(Process.class);
+
+        helper.doStartProcess(transaction, process, "hello");
+        Span span = storageMap.get(process);
+
+        helper.doStartProcess(transaction, process, "hello");
+        assertThat(storageMap.get(process))
+            .describedAs("initial span should not be overwritten")
+            .isSameAs(span);
+    }
+
+    @Test
+    void endTwiceShouldIgnore() {
+        Process process = mock(Process.class);
+
+        helper.doStartProcess(transaction, process, "hello");
+        assertThat(storageMap).isNotEmpty();
+
+        helper.doWaitForEnd(process);
+
+        // this second call should be ignored, thus exception not reported
+        helper.doWaitForEnd(process);
+
+        assertThat(reporter.getSpans()).hasSize(1);
+        assertThat(reporter.getErrors())
+            .describedAs("error should not be reported")
+            .isEmpty();
+    }
+
+    @Test
+    void createMultipleProcessInTransaction() {
```

Comment:
This test does not what the name of the test suggests. There isn't even a assertion.

Reply:
good catch, fixed in the last commit